### PR TITLE
a introduction to the use of redux

### PR DESCRIPTION
**Pull request by Brenden-Yiping-Wang**
This pull request introduces a detailed tutorial on Redux, a prominent state management library in JavaScript. Aimed at developers with a background in JavaScript. The tutorial encompasses a wide range of topics, from the basics of Redux to its core concepts, providing a thorough understanding of its functionality and usage.

### CODE REVIEW

**Comment by Brenden-Yiping-Wang** [comment]
> There are many state managment tools available. Why you choose introduce this one ? What are some of the strength and weakness of it?

The reason why the redux is popular nowadays is because it provides a consistent architecture with strict guidelines, which can make the code more maintainable and easier to understand for new team members. However, there is also a tradeoff. For small to medium-sized projects, the complexity and overhead that Redux introduces may not be necessary, and simpler state management solutions could suffice.

I will make a conclusion in my tutorial and push it again!

**Comment by YuouLei** [comment]
> > There are many state managment tools available. Why you choose introduce this one ? What are some of the strength and weakness of it?
> 
> The reason why the redux is popular nowadays is because it provides a consistent architecture with strict guidelines, which can make the code more maintainable and easier to understand for new team members. However, there is also a tradeoff. For small to medium-sized projects, the complexity and overhead that Redux introduces may not be necessary, and simpler state management solutions could suffice.
> 
> I will make a conclusion in my tutorial and push it again!

Okay, I see. That makes sense. Thanks so much for your explanation.

File: Topics/Tech_Stacks/redux.md [hunk 1]
```diff
@@ -25,6 +26,17 @@ This tutorial provides a comprehensive guide to Redux, a popular state managemen
 
 Redux is a predictable state container for JavaScript applications. It's used for managing the state of your app in a more predictable and easier-to-debug manner. Redux maintains the state of an entire application in a single immutable state tree, which can't be changed directly. When changes are made, a new object is created and merged with the state.
 
+## Why Do We Use Redux?
```
Comment (YuouLei) [inline Topics/Tech_Stacks/redux.md:29]: I see,  Redux maye be is better suited  for large scale applications. Could you explain what efforts Redux make to make the state changes maintainable and easy to understand?

File: Topics/Tech_Stacks/redux.md [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+```bash
+npm install react-redux
+```
+
+## Core Concepts
+
+### Store
+
+The Store is the central hub where all application state is stored. It's created using Redux's `createStore` method. The Store provides several methods:
```
Comment (YuouLei) [inline Topics/Tech_Stacks/redux.md:58]: Could you provide some examples that demonstrate how these function are used? That will definitely help with understanding.

**Review by YuouLei — COMMENTED** [review]
Generally,  some examples will defnitiely help with our understanding. Also, I have some specific question that is attached on the commits.

File: Topics/Tech_Stacks/redux.md (hunk 1 above)
Comment (Brenden-Yiping-Wang) [reply Topics/Tech_Stacks/redux.md:29]: Redux make the state changes maintainable and easy to understand in two ways:
+ Redux maintains the entire state of your application in a single store. This centralized approach ensures that the state is consistent across the app and makes it easier to track and manage state changes.
+ In Redux, the state is immutable, meaning it cannot be changed directly. Instead, any state change produces a new state object. This immutability simplifies tracking changes over time and debugging, as each state in the application's lifecycle can be inspected and understood in isolation.

Comment (YuouLei) [reply Topics/Tech_Stacks/redux.md:29]: That makes sense a lot. So, it basically keep everything is single immutable and trackable state, which enhances the tracking of states across the whole app.

File: Topics/Tech_Stacks/redux.md [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+```
+
+Actions are dispatched using the `dispatch` method of the Store.
+
+### Reducer
+
+Reducers are pure functions that take the current state and an action, and return a new state. They describe how the state changes in response to an action. Reducers should be deterministic, meaning given the same input, they should always return the same output.
+
+A reducer for a to-do application might look like this:
+
```
Comment (YuouLei) [inline Topics/Tech_Stacks/redux.md:122]: Now, I understand that the action is just a javascript object, which represents as the message sending to the store, telling what to update. And The store is like a container that has all the states. However, I don't quite get the concept of reducer. The explanation now might be hard to understand for a person without experience.

**Review by YuouLei — COMMENTED** [review]
Action and Reducer follow up question

File: Topics/Tech_Stacks/redux.md (hunk 2 above)
Comment (Brenden-Yiping-Wang) [reply Topics/Tech_Stacks/redux.md]: Definitely, I will add one example each for the dispatch, getState and dispatch.

File: Topics/Tech_Stacks/redux.md (hunk 3 above)
Comment (Brenden-Yiping-Wang) [reply Topics/Tech_Stacks/redux.md:122]: Generally, in Redux, a reducer is a function that decides how the state of your application changes in response to an action. To put it simply, a reducer acts like a director who takes instructions (actions) and updates the state of the application accordingly. I will add a workflow explaining how to use reducer.

File: Topics/Tech_Stacks/redux.md [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+- Example:
+  ```javascript
+  // Subscribing to state changes
+  const unsubscribe = store.subscribe(() => 
+    console.log('State changed:', store.getState())
+  );
+
+  // To unsubscribe from the state changes
+  unsubscribe();
+  ```
```
Comment (YuouLei) [inline Topics/Tech_Stacks/redux.md:79]: I've grasped the fundamental concepts of Redux, such as the store, action, and reducer. However, managing the state remains a challenge. Could you provide a scenario-based walkthrough, guiding a developer through a practical example of using these concepts in an app? Also, for further learning about Redux, could you recommend any online resources?

**Review by YuouLei — COMMENTED** [review]
Follow up : Scenario walkthorugh

File: Topics/Tech_Stacks/redux.md (hunk 4 above)
Comment (Brenden-Yiping-Wang) [reply Topics/Tech_Stacks/redux.md:79]: You are right, I will add a scenario-based walkthrough to enhance the understanding. Besides, I will also add some websites for further learning.

**Comment by Brenden-Yiping-Wang** [comment]
I add the tutorial for redux, a state management library.

**Pull request merged**